### PR TITLE
Binary search fixes/1

### PR DESCRIPTION
**Pull request by yugr**
@unterumarmung could you review? I've uglified the code but I don't see a better way...

### CODE REVIEW

File: include/sortcheck.h [hunk 1]
```diff
@@ -239,8 +255,14 @@ inline bool binary_search_checked(_ForwardIterator __first,
                                   _ForwardIterator __last,
                                   const _Tp &__val,
                                   const char *file, int line) {
-  Compare<_Tp> compare;
+#ifdef SORTCHECK_SUPPORT_COMPARELESS_API
+  Compare2<SORTCHECK_DECLTYPE(*__first), _Tp> compare;
   return binary_search_checked(__first, __last, __val, compare, file, line);
+#else
+  file = file;
+  line = line;
```
Comment (unterumarmung) [inline include/sortcheck.h]: Is it used to suppress the unused variable warning?
If so, I suggest to introduce such macro:
```cpp
#define SORTCHECK_UNUSED(x) ((void)(x))
```
And then:
```suggestion
  SORTCHECK_UNUSED(file);
  SORTCHECK_UNUSED(line);
```

File: include/sortcheck.h [hunk 2]
```diff
@@ -30,9 +30,20 @@ namespace sortcheck {
 #endif
 
 // TODO: replace with std:less ?
-template<typename T>
-struct Compare {
-  bool operator()(const T &lhs, const T &rhs) const {
+template<typename A>
+struct Compare1 {
+  bool operator()(const A &lhs, const A &rhs) const {
```
Comment (unterumarmung) [inline include/sortcheck.h]: It is not really needed but probably can improve some optimizations
```suggestion
  bool operator()(const A &lhs, const A &rhs) const noexcept(noexcept(lhs < rhs)) {
```
WDYT?

File: tests/bsearch-different-types/repro.cc [hunk 3]
```diff
@@ -4,9 +4,17 @@
...
+#ifdef INVALID
+    "C",
+    "B",
+#else
     "B",
     "C",
+#endif
   };
-  return binary_search(&names[0], &names[sizeof(names) / sizeof(std::string)], name);
+  return std::binary_search(&names[0], &names[sizeof(names) / sizeof(std::string)], name);
```
Comment (unterumarmung) [inline tests/bsearch-different-types/repro.cc:15]:
```suggestion
  return std::binary_search(std::begin(names), std::end(names), name);
```

File: include/sortcheck.h [hunk 4]
```diff
@@ -30,9 +30,20 @@ namespace sortcheck {
 #endif
 
 // TODO: replace with std:less ?
-template<typename T>
-struct Compare {
-  bool operator()(const T &lhs, const T &rhs) const {
+template<typename A>
+struct Compare1 {
```
Comment (unterumarmung) [inline include/sortcheck.h]:
```suggestion
struct SingleTypeLess {
```

File: include/sortcheck.h [hunk 5]
```diff
@@ -30,9 +30,20 @@ namespace sortcheck {
...
+  bool operator()(const A &lhs, const A &rhs) const {
+    return lhs < rhs;
+  }
+};
+
+// TODO: replace with std:less ?
+template<typename A, typename B>
+struct Compare2 {
```
Comment (unterumarmung) [inline include/sortcheck.h]:
```suggestion
struct DifferentTypesLess {
```

File: include/sortcheck.h (hunk 2 above)
Comment (yugr) [reply include/sortcheck.h]: Good idea but will need to wrap it in macro, for C++98 compatibility.

File: tests/bsearch-different-types/repro.cc (hunk 3 above)
Comment (yugr) [reply tests/bsearch-different-types/repro.cc:15]: This code is coming from OSS so I'd rather keep it as is.

**Review by unterumarmung — APPROVED** [review]

**Pull request merged**